// digitaljs/lib/cells/coder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BinDecoder = void 0;

var joint = _interopRequireDefault(require("jointjs"));

var _lodash = _interopRequireDefault(require("lodash"));

var _base = require("./base");

var help = _interopRequireDefault(require("../help"));

var _vl = require("3vl");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const BinDecoder = _base.Box.define('BinDecoder', {
  // default properties
  bits: {
    in: 1,
    out: 2
  },
  constants: undefined
}, {
  initialize() {
    const bits = this.get('bits');
    const constants = this.get('constants') || Array.from(Array(2 ** bits.in).keys()).slice(0, bits.out);
    console.assert(constants.length == bits.out);
    this.get('ports').items = [{
      id: 'in',
      group: 'in',
      dir: 'in',
      bits: bits.in
    }, {
      id: 'out'
    }];

    _base.Box.prototype.initialize.apply(this, arguments);
  },

  _gateParams: _base.Box.prototype._gateParams.concat(['bits', 'constants'])
});

exports.BinDecoder = BinDecoder;
