// digitaljs/lib/tools.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MonitorButton = exports.RemoveButton = exports.DoublyButton = exports.CircleTargetArrowhead = exports.CircleSourceArrowhead = void 0;

var joint = _interopRequireDefault(require("jointjs"));

var _lodash = _interopRequireDefault(require("lodash"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const circleArrowhead = {
  tagName: 'circle',
  attributes: {
    'r': 7,
    'fill': 'black',
    'fill-opacity': 0.3,
    'stroke': 'black',
    'stroke-width': 2,
    'cursor': 'move'
  }
};
const CircleSourceArrowhead = joint.linkTools.SourceArrowhead.extend(_lodash.default.merge({}, circleArrowhead));
exports.CircleSourceArrowhead = CircleSourceArrowhead;
const CircleTargetArrowhead = joint.linkTools.TargetArrowhead.extend(_lodash.default.merge({}, circleArrowhead));
exports.CircleTargetArrowhead = CircleTargetArrowhead;
const DoublyButton = joint.linkTools.Button.extend({
  update() {
    if (this.relatedView.isShortWire()) {
      this.options.distance = this.options.distanceShort || this.options.distance;
      if (this.options.secondary) this.hide();
    } else {
      this.options.distance = this.options.distanceLong || this.options.distance;
    }

    return joint.linkTools.Button.prototype.update.apply(this, arguments);
  }

});
exports.DoublyButton = DoublyButton;
const RemoveButton = DoublyButton.extend({
  name: 'remove',
  children: joint.linkTools.Remove.prototype.children,
  options: joint.linkTools.Remove.prototype.options
});
exports.RemoveButton = RemoveButton;
const MonitorButton = DoublyButton.extend({
  name: 'monitor',
  children: [{
    tagName: 'circle',
    selector: 'button',
    attributes: {
      'r': 7,
      'fill': '#001DFF',
      'cursor': 'pointer'
    }
  }, {
    tagName: 'path',
    selector: 'icon',
    attributes: {
      'd': 'm -2.5,-0.5 a 2,2 0 1 0 4,0 2,2 0 1 0 -4,0 M 1,1 3,3',
      'fill': 'none',
      'stroke': '#FFFFFF',
      'stroke-width': 2,
      'pointer-events': 'none'
    }
  }],
  options: {
    action(evt) {
      this.notify('link:monitor');
    }

  }
});
exports.MonitorButton = MonitorButton;
